e2e tests for federated-apiserver

colhom opened the pull request:
#24880

Here we have federated-apiserver wired up into the e2e testing framework. I've pulled in #24016 and #24943 so that the cluster create/get/delete gingko routine can use a `FederationClient`.

There are a few remaining items to address:
- [x] ~~integrate with `kube-up` such that federated-apiserver deploy option is controlled via an env var. now you have to run `federation/cluster/federated-push.sh` on your own after the build, but before the tests start~~ While the federated binaries/images are built alongside the rest of the components, pushing those images to repository and deploying the federated containers is now a decoupled procedure using scripts in `federation/cluster`.
- [x] make this work with GCE and AWS. (Any other providers we should make sure are in the initial PR?), and error out explicitly for the rest.
- [x] Detect lack of federated-apiserver deploy option and skip federated framework tests automatically.

\cc @nikhiljindal @quinton-hoole

quinton-hoole commented:
cc @kubernetes/goog-testing @fejta @ixdy @lavalamp @mikedanese FYI
This is the stuff to incorporate building and testing Ubernetes Federations of clusters in our continuous integration environment.

quinton-hoole commented on cluster/common.sh:
```diff
@@ -99,8 +99,9 @@ function create-kubeconfig() {
   "${kubectl}" config set-context "${CONTEXT}" --cluster="${CONTEXT}" --user="${CONTEXT}"
-  "${kubectl}" config use-context "${CONTEXT}"  --cluster="${CONTEXT}"
-
+  if [[ "${SECONDARY_KUBECONFIG}" != "true" ]];then
```
What exactly is SECONDARY_KUBECONFIG ? Could you add a comment to explain?

quinton-hoole commented on cluster/kube-util.sh:
```diff
@@ -94,6 +94,83 @@ function test-teardown {
 	echo "TODO: test-teardown" 1>&2
 }
 
+function find-kubectl-binary {
```
It's not clear to my why we need this here? How have the scripts been finding kubectl up to now?

quinton-hoole replied:
Never mind.  I see that all you did is move the existing logic.

lavalamp commented on federation/cluster/template.go:
```diff
@@ -0,0 +1,52 @@
+package main
```
Thank you for doing this instead of writing a hairy bash thing! Can you add a doc commment here?

lavalamp replied:
especially, document the _BASE64 thing.

lavalamp commented on federation/cluster/template.go:
```diff
@@ -0,0 +1,52 @@
+	envList := os.Environ()
+
+	for i := range envList {
+		pieces := eqPattern.Split(envList[i], 2)
```
`strings.SplitN(envList[i], 2)`? Don't need a regexp here, do you?

lavalamp commented on federation/cluster/template.go:
```diff
@@ -0,0 +1,52 @@
+
+func templateYamlFile(params map[string]string, inpath string, out io.Writer) error {
+	if tmpl, err := template.New(path.Base(inpath)).ParseFiles(inpath); err != nil {
+
```
nit: extra blank line. (also line 44)

lavalamp commented on cluster/saltbase/salt/kube-master-addons/kube-master-addons.sh:
```diff
@@ -45,15 +45,24 @@ while true; do
       restart_docker=true
     fi
+
+    timeout 30 docker load -i /srv/salt/kube-bins/federated-apiserver.tar 1>/dev/null 2>&1
```
I am not sure this is an acceptable way to get the container onto nodes. Am thinking about it.

There's two problems with this:
1. it distributes this binary to every master & node, even for users that don't want to run it
2. you can't ship an updated version of the container without tearing the cluster down?

quinton-hoole replied:
+1

quinton-hoole replied:
I would prefer us to create ReplicaSet to run the Federated API server.

lavalamp replied:
You should use a deployment, which creates a replicaset for you.

On Thu, May 5, 2016 at 8:47 AM, Quinton Hoole <notifications@github.com>
wrote:

> In cluster/saltbase/salt/kube-master-addons/kube-master-addons.sh
> <https://github.com/kubernetes/kubernetes/pull/25125#discussion_r62206608>
> :
>
> > @@ -45,15 +45,24 @@ while true; do
> >      elif [[ $rc == 124 ]]; then
> >        restart_docker=true
> >      fi
> > +
> > +    timeout 30 docker load -i /srv/salt/kube-bins/federated-apiserver.tar 1>/dev/null 2>&1
>
> I would prefer us to create ReplicaSet to run the Federated API server.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25125/files/e80616bed4bafaf5f31fc26db9040d843946809d#r62206608>
>

lavalamp replied:
I think it probably does make sense to add a target to the make file, like
`make ubernetes-release` or some such.

On Thu, May 5, 2016 at 11:24 AM, Daniel Smith <dbsmith@google.com> wrote:

> You should use a deployment, which creates a replicaset for you.
>
> On Thu, May 5, 2016 at 8:47 AM, Quinton Hoole <notifications@github.com>
> wrote:
>
>> In cluster/saltbase/salt/kube-master-addons/kube-master-addons.sh
>> <https://github.com/kubernetes/kubernetes/pull/25125#discussion_r62206608>
>> :
>>
>> > @@ -45,15 +45,24 @@ while true; do
>> >      elif [[ $rc == 124 ]]; then
>> >        restart_docker=true
>> >      fi
>> > +
>> > +    timeout 30 docker load -i /srv/salt/kube-bins/federated-apiserver.tar 1>/dev/null 2>&1
>>
>> I would prefer us to create ReplicaSet to run the Federated API server.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/25125/files/e80616bed4bafaf5f31fc26db9040d843946809d#r62206608>
>>
>
>

quinton-hoole commented on cluster/saltbase/salt/kube-master-addons/kube-master-addons.sh:
```diff
@@ -45,15 +45,24 @@ while true; do
+    timeout 30 docker load -i /srv/salt/kube-bins/federated-apiserver.tar 1>/dev/null 2>&1
+    rc=$?
+    if [[ $rc == 0 ]]; then
+      let loadedImageFlags="$loadedImageFlags|8"
```
Please lift these magic constant values (e.g. 8) into named constants.

quinton-hoole commented on cluster/saltbase/salt/kube-master-addons/kube-master-addons.sh:
```diff
@@ -45,15 +45,24 @@ while true; do
+    rc=$?
+    if [[ $rc == 0 ]]; then
+      let loadedImageFlags="$loadedImageFlags|8"
+    elif [[ $rc == 124 ]]; then
```
Same comment as above for 124.

quinton-hoole commented on cluster/saltbase/salt/kube-master-addons/kube-master-addons.sh:
```diff
@@ -45,15 +45,24 @@ while true; do
 
   # required docker images got installed. exit while loop.
-  if [[ $loadedImageFlags == 7 ]]; then break; fi
+  if [[ $loadedImageFlags == 15 ]]; then break; fi
```
Comment as above.

quinton-hoole commented on cluster/saltbase/salt/kube-node-unpacker/kube-node-unpacker.sh:
```diff
@@ -29,10 +29,19 @@ while true; do
+    rc=$?
+    if [[ "${rc}" == 0 ]]; then
+      let loadedImageFlags="${loadedImageFlags}|2"
+    elif [[ "${rc}" == 124 ]]; then
```
Comment as above.

quinton-hoole commented on cluster/saltbase/salt/kube-node-unpacker/kube-node-unpacker.sh:
```diff
@@ -29,10 +29,19 @@ while true; do
 
   # required docker images got installed. exit while loop.
-  if [[ "${loadedImageFlags}" == 1 ]]; then break; fi
+  if [[ "${loadedImageFlags}" == 3 ]]; then break; fi
```
Comment as above.

colhom commented:
@quinton-hoole @lavalamp 

Thanks everyone for the quick review. It's become apparent this is going to need a second pass.

We've discussed this in the sig meeting, and it's become apparent that we need to decouple entirely the provisioning of the federated cluster from the salt nonsense that everyone seems to abhor.

A couple points about the "separation points" to get the discussion started:

1. **Pushing federated images to the cluster**: I was thinking we deploy a docker registry pod/service to the underlying cluster and push federated container images to it. It's entirely decoupled from the salt horror, and I prefer this to requiring that e2e tests push containers to public registries.

2. **Building federated images**: Should we leave this integrated with the existing `make quick-release` system, or extract that process as well into a separate step?

Commit:
Federation build and e2e test integration

Federation components are now buildable and e2e-testable via e2e.go.

colhom commented:
@lavalamp @quinton-hoole I've implemented what we talked about in our last meeting, and updated the checklist at the top of the PR.  To get an idea of what the flow looks like for running e2e tests, check out the [jenkins script](https://github.com/kubernetes/kubernetes/pull/25125/commits/51a6baf6b1bd9701dfb9bbba6808aa43edd4ef1a).

I currently have the entire e2e suite running every test-- depending on what we want, we could ginkgo focus that down to just the federated tests.

I will get this working with gce/gke tomorrow, and also verify that pushing to a gcr.io repo via gcloud is operational (I've been testing with quay.io ). At that point, we should have e2e testing for federated apiserver!

quinton-hoole commented:
Awesome!  Thanks @colhom.  I'll take a closer look in the morning.

quinton-hoole commented:
So it looks like the script only builds a single cluster right now?  Have you given much thought yet to adding the additional two?

colhom commented:
I have e2e tests on GCE working locally. Multiple underlying clusters are provisioned in specific GCE zones, and the federated cluster is deployed to the last of the specified zones. The Gingko tests have access to the kubecontexts for all the underlying clusters, along with the federated kubecontext itself. I have written an [example](https://github.com/colhom/kubernetes/blob/d654d3ffa3922c57437ab018a2121af0d0230689/test/e2e/federated-apiserver.go) e2e tests which registers all the underlying clusters with the federated apiserver, and checks that they're get'able. (\cc @jianhuiz for string format for `Credentials` field)

**Jenkins work**

In the name of expediency, I have duplicated `hack/jenkins/e2e-runner.sh` and introduced the minimal [set of changes](https://github.com/kubernetes/kubernetes/pull/25125/commits/d56fa994250664e7eee5ade83e77a5b3b76497b1) to "make it work" with what we have now. The problem is that the federated logic does not yet "fit" behind `hack/e2e.go`'s CLI interface. Once it does, we'll be able to use `e2e-runner.sh` with little to no changes beyond the environment variables passed in.


I humbly propose we go forward with this sub-optimal approach and file an issue to fit federated e2e provisioning in behind `hack/e2e.go` and delete `federated-e2e-runner.sh`? This essentially involves refactoring `cluster/federation/e2e.sh` into the various `hack/e2e-internal/*` files.

\cc @quinton-hoole @lavalamp

quinton-hoole commented:
Awesome @colhom!
@ixdy @spxtr PTAL

colhom commented:
I have taken another pass at this today, and we have all the functionality behind `hack/e2e.go`. We should now be able to use the same e2e-runner as the regular gce e2e tests!

I have a PR up against test-infra (https://github.com/kubernetes/test-infra/pull/45) which adds our new `gce-federation` e2e job. As we're now using e2e.go for doing this, it's very similar in style to the `gce-ubernetes-lite` job. I need some help setting up the gcr.io repository that we'll push federated images to (see comment on that PR).

Other than that, this **should** all work. I have been unable to test the jenkins job as I'm still finagling the gcloud auth stuff, but otherwise e2e.go has been tested with `FEDERATED=true`.

\cc @spxtr @ixdy @quinton-hoole

spxtr commented on build/common.sh:
```diff
@@ -97,27 +97,31 @@ kube::build::get_docker_wrapped_binaries() {
           kube-apiserver,busybox
           kube-controller-manager,busybox
           kube-scheduler,busybox
+	  federated-apiserver,busybox
```
Use spaces for indentation.

spxtr commented on cluster/common.sh:
```diff
@@ -38,6 +38,10 @@ KUBE_RELEASE_VERSION_REGEX="^v(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*
 KUBE_CI_VERSION_REGEX="^v(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)-(beta|alpha)\\.(0|[1-9][0-9]*)(\\.(0|[1-9][0-9]*)\\+[-0-9a-z]*)?$"
 
+python="python"
+if [[ "`$python --version 2>&1`" != "Python 2."* ]];then
```
Use `$(...)` instead of backticks https://google.github.io/styleguide/shell.xml

colhom commented:
I've been testing this on our K8s Jenkins instance at CoreOS, and everything seems to be working.

Reference kubernetes/test-infra#45 for corresponding Jenkins job definitions. Works with default e2e-runner framework with no changes.

\cc @madhusudancs

madhusudancs commented on build/common.sh:
```diff
@@ -97,27 +97,31 @@ kube::build::get_docker_wrapped_binaries() {
           kube-apiserver,busybox
           kube-controller-manager,busybox
           kube-scheduler,busybox
+          federated-apiserver,busybox
```
Not sure what order pattern this is using, but I think it should either be alphabetical or grouped. I am inclined towards the latter, probably with a comment before the group saying that these are federation binaries. Example:

```shell
# Cluster federation binaries
federated-apiserver,busybox
```

madhusudancs commented on build/common.sh:
```diff
@@ -97,27 +97,31 @@ kube::build::get_docker_wrapped_binaries() {
           kube-apiserver,armel/busybox
           kube-controller-manager,armel/busybox
           kube-scheduler,armel/busybox
+          federated-apiserver,armel/busybox
```
Same comment here and for the two other lists below.

madhusudancs commented on build/util.sh:
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+
+# Copyright 2014 The Kubernetes Authors All rights reserved.
```
Year - 2016

madhusudancs commented on cluster/common.sh:
```diff
@@ -38,6 +38,10 @@ KUBE_RELEASE_VERSION_REGEX="^v(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*
 
+python="python"
+if [[ "$($python --version 2>&1)" != "Python 2."* ]];then
+    python="python2"
```
This is already way better than what we had before. But what happens if python2 doesn't exist?

madhusudancs replied:
Also, the Python code below seems to be Python 3 compatible.

colhom replied:
I think that got introduced in a commit I rebased onto recently. I remember distinctly having to symlink python to python2 :)

madhusudancs commented on cluster/common.sh:
```diff
@@ -50,12 +54,23 @@ KUBE_CI_VERSION_REGEX="^v(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)-(be
+  SECONDARY_KUBECONFIG=${SECONDARY_KUBECONFIG:-false}
+  OVERRIDE_CONTEXT=${OVERRIDE_CONTEXT:-false}
+
+  if [[ "$OVERRIDE_CONTEXT" != "false" ]];then
```
Can't empty string represent false? It still feels strange to store boolean value as string, particularly as nil value.

colhom replied:
Yeah, this is non-conventional. I'm going to change it to be `[[ "$OVERRIDE_CONTEXT" == "true" ]]

madhusudancs commented on cluster/common.sh:
```diff
@@ -99,7 +114,10 @@ function create-kubeconfig() {
   "${kubectl}" config set-context "${CONTEXT}" --cluster="${CONTEXT}" --user="${CONTEXT}"
-  "${kubectl}" config use-context "${CONTEXT}"  --cluster="${CONTEXT}"
+
+  if [[ "${SECONDARY_KUBECONFIG}" != "true" ]];then
```
Is it not possible to do `== "false"` here?

madhusudancs replied:
Also, what happens when `"${SECONDARY_KUBECONFIG}" == "true"`?

colhom replied:
I want the semantics as follows: 

`To skip setting this kubeconfig as the current-context, set SECONDARY_KUBECONFIG=true. Otherwise, the new context will be set as the current-context`

madhusudancs commented on cluster/common.sh:
```diff
@@ -209,7 +227,7 @@ function gen-kube-bearertoken() {
 #   KUBE_UID
 function gen-uid {
-    KUBE_UID=$(python -c 'import uuid; print(uuid.uuid1().fields[0])')
+    KUBE_UID=$($python -c 'import uuid; print uuid.uuid1().fields[0]')
```
Why make this Python 3 incompatible?

madhusudancs commented on cluster/kube-util.sh:
```diff
@@ -99,3 +99,32 @@ PROVIDER_UTILS="${KUBE_ROOT}/cluster/${KUBERNETES_PROVIDER}/util.sh"
     source "${PROVIDER_UTILS}"
 fi
+
+# Federated utils
```
"Federation utils" I guess?

nikhiljindal replied:
yes +1

madhusudancs commented on cluster/kube-util.sh:
```diff
@@ -99,3 +99,32 @@ PROVIDER_UTILS="${KUBE_ROOT}/cluster/${KUBERNETES_PROVIDER}/util.sh"
+
+# Federated utils
+
+# Should generally not be called within the global scope
```
It is worth explaining the exceptions.

nikhiljindal commented on federation/manifests/federation-ns.yaml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: {{.FEDERATED_NAMESPACE}}
```
How are these env vars getting replaced by their values?

colhom replied:
[templating routine](https://github.com/colhom/kubernetes/blob/federation-e2e/federation/cluster/template.go), [example usage](https://github.com/colhom/kubernetes/blob/federation-e2e/federation/cluster/common.sh#L57)

nikhiljindal replied:
hmm. So how do I use these manifests to bring up the federation control plane (for developing or setting up federation)?
I was hoping I can just do `kubectl create -f federation/manifests` but clearly that is not going to be enough. I guess we will need a script that will first replace the env vars by their values and then call kubectl create?

colhom replied:
yep. the alternative is that we create a `federation/manifests/default/` folder which contains the results of templating the files with "default values".

nikhiljindal commented on federation/manifests/federated-apiserver-deployment.yaml:
```diff
@@ -0,0 +1,43 @@
+kind: Deployment
+metadata:
+  name: {{.FEDERATED_APISERVER_DEPLOYMENT_NAME}}
+  namespace: federation-e2e
```
Should this be {{.FEDERATED_NAMESPACE}}?

colhom replied:
Yep!

nikhiljindal commented on federation/manifests/federated-apiserver-deployment.yaml:
```diff
@@ -0,0 +1,43 @@
+  name: {{.FEDERATED_APISERVER_DEPLOYMENT_NAME}}
+  namespace: federation-e2e
+  labels:
+    app: federation-e2e
```
I was hoping that we can keep these manifests generic, so that I can use them to bring up a local federation control place or can use them in e2e. So will be great if we can get rid of all the e2e references in these yaml files.

colhom replied:
Agreed.

nikhiljindal commented on federation/manifests/federated-apiserver-deployment.yaml:
```diff
@@ -0,0 +1,43 @@
+        - --secure-port=443
+        - --advertise-address={{.FEDERATED_API_HOST}}
+        - --admission-control=NamespaceLifecycle,LimitRanger,SecurityContextDeny,ServiceAccount,ResourceQuota
+        - --runtime-config=federation/v1alpha1
```
No need to pass this. This is only for overrides.

nikhiljindal commented on test/e2e/federated-apiserver.go:
```diff
@@ -0,0 +1,65 @@
+						},
+					},
+					//TODO(colhom): took a guess at the format. is this right?
+					Credential: fmt.Sprintf("basicauth:%s:%s", context.User.User.Username, context.User.User.Password),
```
Fwiw, cluster controller is not using this field.
Ref https://github.com/kubernetes/kubernetes/issues/25761

colhom replied:
So this field is fine to omit from the struct definition?

nikhiljindal replied:
Yes. For now just omit it.
I am replacing it by a secretRef in https://github.com/kubernetes/kubernetes/pull/26132

nikhiljindal commented:
@colhom Do you think you can fix the comments and get this merged before the merge freeze today?

nikhiljindal commented on federation/cluster/template.go:
```diff
@@ -0,0 +1,74 @@
+)
+
+/*
+This is a simple script that make *every* environment variable available
```
make -> makes

nikhiljindal commented on federation/cluster/common.sh:
```diff
@@ -0,0 +1,189 @@
+: "${KUBE_ROOT?Must set KUBE_ROOT env var}"
+
+FEDERATED_IMAGE_REPO_BASE=${FEDERATED_IMAGE_REPO_BASE:-'gcr.io/google_containers'}
+FEDERATED_NAMESPACE=${FEDERATED_NAMESPACE:-federation-e2e}
```
Replace FEDERATED by FEDERATION ex: FEDERATION_NAMESPACE

nikhiljindal replied:
everywhere

colhom replied:
Done

nikhiljindal commented on hack/lib/golang.sh:
```diff
@@ -162,6 +163,7 @@ readonly KUBE_STATIC_LIBRARIES=(
   kube-scheduler
   kube-proxy
   kubectl
+  federated-apiserver
```
FYI, I am renaming this to federation-apiserver in https://github.com/kubernetes/kubernetes/pull/26196

nikhiljindal commented on test/e2e/framework/framework.go:
```diff
@@ -37,8 +38,11 @@ import (
 	"k8s.io/kubernetes/pkg/util/intstr"
 	"k8s.io/kubernetes/pkg/util/wait"
 
+	"io/ioutil"
```
Put this at the top (along with fmt, strings, etc)

colhom replied:
Done

nikhiljindal commented on test/e2e/framework/framework.go:
```diff
@@ -54,6 +58,8 @@ type Framework struct {
 	Clientset_1_2 *release_1_2.Clientset
 	Clientset_1_3 *release_1_3.Clientset
 
+	FederatedClient *unversionedfederation.FederationClient
```
FederatedClient -> FederationClient

colhom replied:
Done.

nikhiljindal commented on test/e2e/kubectl.go:
```diff
@@ -38,9 +38,9 @@ import (
 	"time"
 
 	"github.com/ghodss/yaml"
+	"k8s.io/kubernetes/pkg/api/annotations"
```
I assume this is unintended. Revert

colhom replied:
This has been reverted

colhom commented on cluster/common.sh:
```diff
@@ -222,7 +222,7 @@ function gen-kube-bearertoken() {
 #   KUBE_UID
 function gen-uid {
-    KUBE_UID=$(python -c 'import uuid; print uuid.uuid1().fields[0]')
+    KUBE_UID=$(python -c 'import uuid; print(uuid.uuid1().fields[0])')
```
this was a revert to master.

colhom commented on test/e2e/kubectl.go:
```diff
@@ -40,9 +40,9 @@ import (
 
 	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/annotations"
 	apierrs "k8s.io/kubernetes/pkg/api/errors"
```
this is a revert to master

colhom commented:
@nikhiljindal I've addressed your review comments. Corresponding changes to test-infra [here](https://github.com/kubernetes/test-infra/pull/45/commits/79187772f2e831767447b24ec024a2eca6cf28c9)

Once this review cycle is complete, I'll run this through Jenkins over here and verify everything works.

We still have a bit of `federated --> federation` renaming to contend with, but we can sync that up later with your PR which renames the apiserver.

nikhiljindal commented on cluster/common.sh:
```diff
@@ -99,7 +109,10 @@ function create-kubeconfig() {
   "${kubectl}" config set-context "${CONTEXT}" --cluster="${CONTEXT}" --user="${CONTEXT}"
-  "${kubectl}" config use-context "${CONTEXT}"  --cluster="${CONTEXT}"
+
+  if [[ "${SECONDARY_KUBECONFIG}" != "" ]];then
```
Sorry maybe I am not understanding this correctly, if SECONDARY_KUBECONFIG = true, the following command of setting the context should not be run, right?
Thats not what this condition will do

colhom replied:
That's correct, my mistake. I'm going to change it to ` != "true"`. To bypass setting this context as current, set `SECONDARY_KUBECONFIG=true`.

nikhiljindal commented:
Would be great if @quinton-hoole or someone from @kubernetes/test-infra-maintainers takes a look as well.

Also, please feel free to squash your commits.

nikhiljindal commented:
And remove WIP from title now :)

colhom commented:
@nikhiljindal squashed and running through jenkins over here.

nikhiljindal commented:
Great, thanks!

Adding the LGTM label.
Once this is merged, multiple people can add e2e test cases in parallel.

nikhiljindal commented:
My PR https://github.com/kubernetes/kubernetes/pull/26196 got merged, so you will have to update the references to federated-apiserver (change it to federation-apiserver).

Removing LGTM.

colhom commented:
@nikhiljindal - jenkins run was a success!

```sh
22:32:49 Run starting at Tue May 24 22:32:48 UTC 2016
22:32:49 Found Kubernetes version: v1.3.0-alpha.0.3316+dfe9a5a19004ac

....

22:48:22 [1m[32mRan 1 of 301 Specs in 18.728 seconds[0m
22:48:22 [1m[32mSUCCESS![0m -- [32m[1m1 Passed[0m | [91m[1m0 Failed[0m | [33m[1m0 Pending[0m | [36m[1m300 Skipped[0m PASS
22:48:22 
22:48:22 Ginkgo ran 1 suite in 19.32844064s
22:48:22 Test Suite Passed
```

colhom commented:
@nikhiljindal I've rebased over your changes and have done some more sed wrangling on the files to reflect your rename PR. I have the Jenkins loop running again, and will squash and ping you when it finishes in about an hour.

colhom commented:
@nikhiljindal Jenkins is happy with the renaming! Let me know when you've reviewed that last commit, I'll squash up and you can drop the lgtm again.

nikhiljindal commented:
Yes. Looks good. thanks!

Also, can you specify here the commands I need to run to run the federation tests?
For ex: today, I need to run the following commands to run the k8s tests:
```
$ go run hack/e2e.go -v --up # to bring up the test cluster
$ go run hack/e2e.go -v -test --test_args='--ginkgo.focus=.*LimitRange.*' # Run this command multiple times to run the LimitRange test
```
Is just replacing LimitRange by Federation in second command enough?

colhom commented:
@nikhiljindal asked how to run this

```sh
# ASSUME KUBERNETES_PROVIDER=gce and otherwise working e2e setup already
$ export FEDERATION=true
$ export E2E_ZONES="us-central1-a us-central1-b us-central1-f"
$ export FEDERATION_PUSH_REPO_BASE="gcr.io/your-gcloud-project-name"

# Due to hackiness, the main "zone" is always the last in the E2E_ZONES list
$ export KUBE_GCE_ZONE="us-central1-f"

# The build step pushes the federation container images to gcr
$ go run hack/e2e/go -v --build

$ go run hack/e2e.go -v --up
$ go run hack/e2e.go -v -test --test_args="--ginkgo.focus=\[Feature:Performance\]"

colhom commented:
@nikhiljindal squashed up.

nikhiljindal commented:
LGTM, thanks :)

nikhiljindal commented:
Also FYI, in https://github.com/kubernetes/kubernetes/pull/26132 (not merged yet) I am using POD_NAMESPACE env var in ClusterController to figure out the namespace that federation-controller-manager is running in. I need the namespace to fetch the secret containing kubeconfigs for k8s clusters.

We will need to provide that env var using Downward API, so we need to add the following to spec.containers in deployment.yaml for kube-controller-manager deployment:

```
env:
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
```

You dont necessarily have to add that in this PR (you are free to :) or I will send a follow up PR), but I wanted to let you know that we will require that change to your manifest file.

colhom commented:
@nikhiljindal we had a integration test failure

```sh
Verifying ./hack/../hack/verify-flags-underscore.py
Found flags in golang files not in the list of known flags. Please add these to hack/verify-flags/known-flags.txt
federated-kube-context
kube-context
FAILED   ./hack/../hack/verify-flags-underscore.py	6s
Build step 'Execute shell' marked build as failure
```

I've patched and took the liberty of squashing. Here's the diff on what you lgtm'd

```sh
$ git diff origin/federation-e2e
diff --git a/hack/e2e.go b/hack/e2e.go
index dff7427..ee2b338 100644
--- a/hack/e2e.go
+++ b/hack/e2e.go
@@ -49,8 +49,7 @@ var (
                "You can explicitly set to false if you're, e.g., testing single-node clusters "+
                "for which the node count is supposed to be one.")
 
-       ctlCmd      = flag.String("ctl", "", "If nonempty, pass this as an argument, and call kubectl. Implies -v. (-test, -cfg, -ctl are mutually exclusive)")
-       kubeContext = flag.String("kube-context", "", "If nonempty, force use of context. Otherwise use current-context")
+       ctlCmd = flag.String("ctl", "", "If nonempty, pass this as an argument, and call kubectl. Implies -v. (-test, -cfg, -ctl are mutually exclusive)")
 )
 
 const (
diff --git a/hack/verify-flags/known-flags.txt b/hack/verify-flags/known-flags.txt
index e485152..cae47d5 100644
--- a/hack/verify-flags/known-flags.txt
+++ b/hack/verify-flags/known-flags.txt
@@ -143,6 +143,7 @@ failure-domains
 fake-clientset
 federated-api-burst
 federated-api-qps
+federated-kube-context
 file-check-frequency
 file-suffix
 file_content_in_loop
```

nikhiljindal commented:
Thanks. Now its failing with the following:

```
Verifying ./hack/../hack/verify-boilerplate.sh
Boilerplate header is wrong for: /go/src/k8s.io/kubernetes/federation/cluster/template.go
FAILED   ./hack/../hack/verify-boilerplate.sh	1s
```

colhom commented:
Alright, I've patched and squashed again. Moved the header and description to the top of the file and updated the date. Here's the diff:

```sh
$ git diff origin/federation-e2e
diff --git a/federation/cluster/template.go b/federation/cluster/template.go
index 0d6a395..9c947bc 100644
--- a/federation/cluster/template.go
+++ b/federation/cluster/template.go
@@ -1,29 +1,18 @@
-package main
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
 
-// Copyright 2014 The Kubernetes Authors All rights reserved.
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//     http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
 
-import (
-       "encoding/base64"
-       "flag"
-       "fmt"
-       "io"
-       "os"
-       "path"
-       "strings"
-       "text/template"
-)
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
 
 /*
 This is a simple script that makes *every* environment variable available
@@ -35,6 +24,20 @@ $ MYVAR=foobar go run template.go test.txt
 
 If you want the base64 version of any MYVAR, simple use {{.MYVAR_BASE64}}
 */
+
+package main
+
+import (
+       "encoding/base64"
+       "flag"
+       "fmt"
+       "io"
+       "os"
+       "path"
+       "strings"
+       "text/template"
+)
+
 func main() {
        flag.Parse()
        env := make(map[string]string)
```

nikhiljindal commented:
Still failing:

```
Verifying ./hack/../hack/verify-gofmt.sh
!!! 'gofmt -s' needs to be run on the following files: 
./test/e2e/framework/framework.go
FAILED   ./hack/../hack/verify-gofmt.sh	9s
Verifying ./hack/../hack/verify-govet.sh
test/e2e/framework/framework.go:505: struct field tag `yaml:token"` not compatible with reflect.StructTag.Get: bad syntax for struct tag value
FAILED   ./hack/../hack/verify-govet.sh	103s
```

nikhiljindal commented:
Btw, you can run `hack/verify-all.sh` locally to verify that all these verify scripts are satisfied.

colhom commented:
@nikhiljindal GCE e2e is failing:
```sh
ERROR: (gcloud.compute.networks.create) Some requests did not succeed:
 - Quota 'NETWORKS' exceeded. Limit: 75.0
```

colhom commented:
Also, the integration test is failing, but it looks like a flake this time:

```sh
Error: No such image or container: kube-build-data-384106b65e
```

nikhiljindal commented:
Quota issue is https://github.com/kubernetes/kubernetes/issues/26270

nikhiljindal commented:
@k8s-bot e2e test this issue: #26270

nikhiljindal commented:
Unit test failure is still:

```
Verifying ./hack/../hack/verify-gofmt.sh
!!! 'gofmt -s' needs to be run on the following files: 
./test/e2e/framework/framework.go
FAILED   ./hack/../hack/verify-gofmt.sh	16s
```

colhom commented:
@nikhiljindal we're good to go!

nikhiljindal commented:
Woohoo!!

LGTM, thanks!

alex-mohr commented:
Submit queue is blocked on kubemark so emulating a manual submit queue because we suspect we know why it's blocked.
@k8s-bot test this please issue #IGNORE

alex-mohr commented:
buildcop manually merging due to known issue with CI tests

nikhiljindal commented:
Thanks @alex-mohr !

nikhiljindal commented:
fwiw, when I am trying this locally, up never succeeds for me.

Here are the logs:

```
$ FEDERATION=true E2E_ZONES="us-central1-a us-central1-b us-central1-f" KUBE_GCE_ZONE="us-central1-f" go run hack/e2e.go -v --up

2016/05/31 16:16:33 e2e.go:212: Running: get status
Your active configuration is: [default]

Project: pelagic-height-90321
Zone: us-central1-f
Setting zone vars to: federation-e2e-gce-us-central1-a

	Checking version for federation-e2e-gce-us-central1-a
Client Version: version.Info{Major:"1", Minor:"3+", GitVersion:"v1.3.0-alpha.4.833+8d74decee6ae1e-dirty", GitCommit:"8d74decee6ae1e9fc9e43dc5a7c8c5e85ebf68c8", GitTreeState:"dirty", BuildDate:"2016-05-31T22:48:31Z", GoVersion:"go1.6.1", Compiler:"gc", Platform:"linux/amd64"}
error: couldn't read version from server: Get https://146.148.50.86/version: dial tcp 146.148.50.86:443: i/o timeout
2016/05/31 16:17:09 e2e.go:218: Error running get status: exit status 1
2016/05/31 16:17:09 e2e.go:214: Step 'get status' finished in 36.123450892s
2016/05/31 16:17:09 e2e.go:212: Running: up
Your active configuration is: [default]

Project: pelagic-height-90321
Zone: us-central1-f
Setting zone vars to: federation-e2e-gce-us-central1-a
... Starting cluster in us-central1-a using provider gce
... calling verify-prereqs
WARNING: Additional component repositories are currently active.  Running `update` instead of `install`.
WARNING: You are using additional component repository: [https://storage.googleapis.com/endpoints-gcloud/components-2.json]
ERROR: (gcloud.components.install) Failed to fetch component listing from server. Check your network settings and try again.
Please ensure that the additional component repository [https://storage.googleapis.com/endpoints-gcloud/components-2.json] is correct and still valid.  To remove it, run:
  $ gcloud components repositories remove https://storage.googleapis.com/endpoints-gcloud/components-2.json 
WARNING: Additional component repositories are currently active.  Running `update` instead of `install`.
WARNING: You are using additional component repository: [https://storage.googleapis.com/endpoints-gcloud/components-2.json]
ERROR: (gcloud.components.install) Failed to fetch component listing from server. Check your network settings and try again.
Please ensure that the additional component repository [https://storage.googleapis.com/endpoints-gcloud/components-2.json] is correct and still valid.  To remove it, run:
  $ gcloud components repositories remove https://storage.googleapis.com/endpoints-gcloud/components-2.json 
WARNING: You are using additional component repository: [https://storage.googleapis.com/endpoints-gcloud/components-2.json]
ERROR: (gcloud.components.update) Failed to fetch component listing from server. Check your network settings and try again.
Please ensure that the additional component repository [https://storage.googleapis.com/endpoints-gcloud/components-2.json] is correct and still valid.  To remove it, run:
  $ gcloud components repositories remove https://storage.googleapis.com/endpoints-gcloud/components-2.json 
... calling kube-up
Your active configuration is: [default]

Project: pelagic-height-90321
Zone: us-central1-a
gs://kubernetes-staging-10882c4b98/devel/
gs://kubernetes-staging-10882c4b98/e2e-test-nikhiljindal-devel/
gs://kubernetes-staging-10882c4b98/nikhiljindal-us-central1-a-devel/
gs://kubernetes-staging-10882c4b98/nikhiljindal-us-central1-b-devel/
gs://kubernetes-staging-10882c4b98/nikhiljindal-us-central1-f-devel/
+++ Staging server tars to Google Storage: gs://kubernetes-staging-10882c4b98/nikhiljindal-us-central1-a-devel
+++ kubernetes-server-linux-amd64.tar.gz uploaded (sha1 = 35eb03b5e5be0c974f01287a5f917560009a50c7)
+++ kubernetes-salt.tar.gz uploaded (sha1 = 318fe207033eb2bd3809a92ee991cb91e8f3f8a0)
+++ kubernetes-manifests.tar.gz uploaded (sha1 = e25dc1044c890412ead8629cdf61aeb4065223f7)
Listed 0 items.
INSTANCE_GROUPS=
NODE_NAMES=
Looking for already existing resources
Listed 0 items.
Listed 0 items.
Starting master and configuring firewalls
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/zones/us-central1-a/disks/nikhiljindal-us-central1-a-master-pd].
NAME                                 ZONE          SIZE_GB TYPE   STATUS
nikhiljindal-us-central1-a-master-pd us-central1-a 20      pd-ssd READY
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/regions/us-central1/addresses/nikhiljindal-us-central1-a-master-ip].
Generating certs for alternate-names: IP:104.197.140.83,IP:10.0.0.1,DNS:kubernetes,DNS:kubernetes.default,DNS:kubernetes.default.svc,DNS:kubernetes.default.svc.cluster.local,DNS:nikhiljindal-us-central1-a-master
+++ Logging using Fluentd to elasticsearch
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/firewalls/nikhiljindal-us-central1-a-master-https].
NAME                                    NETWORK SRC_RANGES RULES   SRC_TAGS TARGET_TAGS
nikhiljindal-us-central1-a-master-https e2e     0.0.0.0/0  tcp:443          nikhiljindal-us-central1-a-master
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/zones/us-central1-a/instances/nikhiljindal-us-central1-a-master].
NAME                              ZONE          MACHINE_TYPE  PREEMPTIBLE INTERNAL_IP EXTERNAL_IP    STATUS
nikhiljindal-us-central1-a-master us-central1-a n1-standard-1             10.240.0.5  104.197.140.83 RUNNING
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/firewalls/nikhiljindal-us-central1-a-minion-all].
NAME                                  NETWORK SRC_RANGES    RULES                    SRC_TAGS TARGET_TAGS
nikhiljindal-us-central1-a-minion-all e2e     10.180.0.0/14 tcp,udp,icmp,esp,ah,sctp          nikhiljindal-us-central1-a-minion
Creating minions.
Attempt 1 to create nikhiljindal-us-central1-a-minion-template
WARNING: You have selected a disk size of under [200GB]. This may result in poor I/O performance. For more information, see: https://developers.google.com/compute/docs/disks/persistent-disks#pdperformance.
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/instanceTemplates/nikhiljindal-us-central1-a-minion-template].
NAME                                       MACHINE_TYPE  PREEMPTIBLE CREATION_TIMESTAMP
nikhiljindal-us-central1-a-minion-template n1-standard-2             2016-05-31T16:18:55.094-07:00
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/zones/us-central1-a/instanceGroupManagers/nikhiljindal-us-central1-a-minion-group].
NAME                                    ZONE          BASE_INSTANCE_NAME                      SIZE TARGET_SIZE INSTANCE_TEMPLATE                          AUTOSCALED
nikhiljindal-us-central1-a-minion-group us-central1-a nikhiljindal-us-central1-a-minion-group      3           nikhiljindal-us-central1-a-minion-template
Waiting for group to become stable, current operations: creating: 3
Waiting for group to become stable, current operations: creating: 3
Waiting for group to become stable, current operations: creating: 3
Waiting for group to become stable, current operations: creating: 3
Group is stable
INSTANCE_GROUPS=nikhiljindal-us-central1-a-minion-group
NODE_NAMES=nikhiljindal-us-central1-a-minion-group-ddfr nikhiljindal-us-central1-a-minion-group-l56h nikhiljindal-us-central1-a-minion-group-q4cj
Using master: nikhiljindal-us-central1-a-master (external IP: 104.197.140.83)
Waiting up to 300 seconds for cluster initialization.

  This will continually check to see if the API for kubernetes is reachable.
  This may time out if there was some uncaught error during start up.

...............................Kubernetes cluster created.
cluster "federation-e2e-gce-us-central1-a" set.
user "federation-e2e-gce-us-central1-a" set.
context "federation-e2e-gce-us-central1-a" set.
switched to context "federation-e2e-gce-us-central1-a".
user "federation-e2e-gce-us-central1-a-basic-auth" set.
Wrote config for federation-e2e-gce-us-central1-a to /usr/local/google/home/nikhiljindal/.kube/config

Kubernetes cluster is running.  The master is running at:

  https://104.197.140.83

The user name and password to use is located in /usr/local/google/home/nikhiljindal/.kube/config.

... calling validate-cluster
Waiting for 4 ready nodes. 1 ready nodes, 4 registered. Retrying.
Waiting for 4 ready nodes. 1 ready nodes, 4 registered. Retrying.
Waiting for 4 ready nodes. 1 ready nodes, 4 registered. Retrying.
Found 4 node(s).
NAME                                           STATUS                     AGE
nikhiljindal-us-central1-a-master              Ready,SchedulingDisabled   1m
nikhiljindal-us-central1-a-minion-group-ddfr   Ready                      1m
nikhiljindal-us-central1-a-minion-group-l56h   Ready                      1m
nikhiljindal-us-central1-a-minion-group-q4cj   Ready                      1m
Validate output:
NAME                 STATUS    MESSAGE              ERROR
scheduler            Healthy   ok                   
controller-manager   Healthy   ok                   
etcd-0               Healthy   {"health": "true"}   
etcd-1               Healthy   {"health": "true"}   
Cluster validation succeeded
Done, listing cluster services:

Kubernetes master is running at https://104.197.140.83
GLBCDefaultBackend is running at https://104.197.140.83/api/v1/proxy/namespaces/kube-system/services/default-http-backend
Elasticsearch is running at https://104.197.140.83/api/v1/proxy/namespaces/kube-system/services/elasticsearch-logging
Heapster is running at https://104.197.140.83/api/v1/proxy/namespaces/kube-system/services/heapster
Kibana is running at https://104.197.140.83/api/v1/proxy/namespaces/kube-system/services/kibana-logging
KubeDNS is running at https://104.197.140.83/api/v1/proxy/namespaces/kube-system/services/kube-dns
kubernetes-dashboard is running at https://104.197.140.83/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard
Grafana is running at https://104.197.140.83/api/v1/proxy/namespaces/kube-system/services/monitoring-grafana
InfluxDB is running at https://104.197.140.83/api/v1/proxy/namespaces/kube-system/services/monitoring-influxdb

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.

NAME                                                        NETWORK SRC_RANGES RULES           SRC_TAGS TARGET_TAGS
e2e-test-nikhiljindal-minion-e2e-test-nikhiljindal-http-alt e2e     0.0.0.0/0  tcp:80,tcp:8080          e2e-test-nikhiljindal-minion
allowed:
- IPProtocol: tcp
  ports:
  - '80'
- IPProtocol: tcp
  ports:
  - '8080'
creationTimestamp: '2016-05-31T16:22:34.544-07:00'
description: ''
id: '7388234897017774677'
kind: compute#firewall
name: e2e-test-nikhiljindal-minion-e2e-test-nikhiljindal-http-alt
network: https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/networks/e2e
selfLink: https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/firewalls/e2e-test-nikhiljindal-minion-e2e-test-nikhiljindal-http-alt
sourceRanges:
- 0.0.0.0/0
targetTags:
- e2e-test-nikhiljindal-minion
NAME                                                         NETWORK SRC_RANGES RULES                           SRC_TAGS TARGET_TAGS
e2e-test-nikhiljindal-minion-e2e-test-nikhiljindal-nodeports e2e     0.0.0.0/0  tcp:30000-32767,udp:30000-32767          e2e-test-nikhiljindal-minion
allowed:
- IPProtocol: tcp
  ports:
  - 30000-32767
- IPProtocol: udp
  ports:
  - 30000-32767
creationTimestamp: '2016-05-31T16:23:03.389-07:00'
description: ''
id: '7482902251168834104'
kind: compute#firewall
name: e2e-test-nikhiljindal-minion-e2e-test-nikhiljindal-nodeports
network: https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/networks/e2e
selfLink: https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/firewalls/e2e-test-nikhiljindal-minion-e2e-test-nikhiljindal-nodeports
sourceRanges:
- 0.0.0.0/0
targetTags:
- e2e-test-nikhiljindal-minion
Setting zone vars to: federation-e2e-gce-us-central1-b
... Starting cluster in us-central1-b using provider gce
... calling verify-prereqs
WARNING: Additional component repositories are currently active.  Running `update` instead of `install`.
WARNING: You are using additional component repository: [https://storage.googleapis.com/endpoints-gcloud/components-2.json]
ERROR: (gcloud.components.install) Failed to fetch component listing from server. Check your network settings and try again.
Please ensure that the additional component repository [https://storage.googleapis.com/endpoints-gcloud/components-2.json] is correct and still valid.  To remove it, run:
  $ gcloud components repositories remove https://storage.googleapis.com/endpoints-gcloud/components-2.json 
WARNING: Additional component repositories are currently active.  Running `update` instead of `install`.
WARNING: You are using additional component repository: [https://storage.googleapis.com/endpoints-gcloud/components-2.json]
ERROR: (gcloud.components.install) Failed to fetch component listing from server. Check your network settings and try again.
Please ensure that the additional component repository [https://storage.googleapis.com/endpoints-gcloud/components-2.json] is correct and still valid.  To remove it, run:
  $ gcloud components repositories remove https://storage.googleapis.com/endpoints-gcloud/components-2.json 
WARNING: You are using additional component repository: [https://storage.googleapis.com/endpoints-gcloud/components-2.json]
ERROR: (gcloud.components.update) Failed to fetch component listing from server. Check your network settings and try again.
Please ensure that the additional component repository [https://storage.googleapis.com/endpoints-gcloud/components-2.json] is correct and still valid.  To remove it, run:
  $ gcloud components repositories remove https://storage.googleapis.com/endpoints-gcloud/components-2.json 
... calling kube-up
Your active configuration is: [default]

Project: pelagic-height-90321
Zone: us-central1-b
gs://kubernetes-staging-10882c4b98/devel/
gs://kubernetes-staging-10882c4b98/e2e-test-nikhiljindal-devel/
gs://kubernetes-staging-10882c4b98/nikhiljindal-us-central1-a-devel/
gs://kubernetes-staging-10882c4b98/nikhiljindal-us-central1-b-devel/
gs://kubernetes-staging-10882c4b98/nikhiljindal-us-central1-f-devel/
+++ Staging server tars to Google Storage: gs://kubernetes-staging-10882c4b98/nikhiljindal-us-central1-b-devel
+++ kubernetes-server-linux-amd64.tar.gz uploaded (sha1 = 35eb03b5e5be0c974f01287a5f917560009a50c7)
+++ kubernetes-salt.tar.gz uploaded (sha1 = 318fe207033eb2bd3809a92ee991cb91e8f3f8a0)
+++ kubernetes-manifests.tar.gz uploaded (sha1 = e25dc1044c890412ead8629cdf61aeb4065223f7)
Listed 0 items.
INSTANCE_GROUPS=
NODE_NAMES=
Looking for already existing resources
Listed 0 items.
Listed 0 items.
Starting master and configuring firewalls
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/zones/us-central1-b/disks/nikhiljindal-us-central1-b-master-pd].
NAME                                 ZONE          SIZE_GB TYPE   STATUS
nikhiljindal-us-central1-b-master-pd us-central1-b 20      pd-ssd READY
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/regions/us-central1/addresses/nikhiljindal-us-central1-b-master-ip].
Generating certs for alternate-names: IP:104.197.226.104,IP:10.0.0.1,DNS:kubernetes,DNS:kubernetes.default,DNS:kubernetes.default.svc,DNS:kubernetes.default.svc.cluster.local,DNS:nikhiljindal-us-central1-b-master
+++ Logging using Fluentd to elasticsearch
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/firewalls/nikhiljindal-us-central1-b-master-https].
NAME                                    NETWORK SRC_RANGES RULES   SRC_TAGS TARGET_TAGS
nikhiljindal-us-central1-b-master-https e2e     0.0.0.0/0  tcp:443          nikhiljindal-us-central1-b-master
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/firewalls/nikhiljindal-us-central1-b-minion-all].
NAME                                  NETWORK SRC_RANGES    RULES                    SRC_TAGS TARGET_TAGS
nikhiljindal-us-central1-b-minion-all e2e     10.180.0.0/14 tcp,udp,icmp,esp,ah,sctp          nikhiljindal-us-central1-b-minion
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/zones/us-central1-b/instances/nikhiljindal-us-central1-b-master].
NAME                              ZONE          MACHINE_TYPE  PREEMPTIBLE INTERNAL_IP EXTERNAL_IP     STATUS
nikhiljindal-us-central1-b-master us-central1-b n1-standard-1             10.240.0.9  104.197.226.104 RUNNING
Creating minions.
Attempt 1 to create nikhiljindal-us-central1-b-minion-template
WARNING: You have selected a disk size of under [200GB]. This may result in poor I/O performance. For more information, see: https://developers.google.com/compute/docs/disks/persistent-disks#pdperformance.
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/instanceTemplates/nikhiljindal-us-central1-b-minion-template].
NAME                                       MACHINE_TYPE  PREEMPTIBLE CREATION_TIMESTAMP
nikhiljindal-us-central1-b-minion-template n1-standard-2             2016-05-31T16:25:11.122-07:00
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/zones/us-central1-b/instanceGroupManagers/nikhiljindal-us-central1-b-minion-group].
NAME                                    ZONE          BASE_INSTANCE_NAME                      SIZE TARGET_SIZE INSTANCE_TEMPLATE                          AUTOSCALED
nikhiljindal-us-central1-b-minion-group us-central1-b nikhiljindal-us-central1-b-minion-group      3           nikhiljindal-us-central1-b-minion-template
Waiting for group to become stable, current operations: creating: 3
Waiting for group to become stable, current operations: creating: 3
Waiting for group to become stable, current operations: creating: 3
Waiting for group to become stable, current operations: creating: 3
Group is stable
INSTANCE_GROUPS=nikhiljindal-us-central1-b-minion-group
NODE_NAMES=nikhiljindal-us-central1-b-minion-group-4f5s nikhiljindal-us-central1-b-minion-group-e8eg nikhiljindal-us-central1-b-minion-group-yxmf
Using master: nikhiljindal-us-central1-b-master (external IP: 104.197.226.104)
Waiting up to 300 seconds for cluster initialization.

  This will continually check to see if the API for kubernetes is reachable.
  This may time out if there was some uncaught error during start up.

........................................Kubernetes cluster created.
cluster "federation-e2e-gce-us-central1-b" set.
user "federation-e2e-gce-us-central1-b" set.
context "federation-e2e-gce-us-central1-b" set.
switched to context "federation-e2e-gce-us-central1-b".
user "federation-e2e-gce-us-central1-b-basic-auth" set.
Wrote config for federation-e2e-gce-us-central1-b to /usr/local/google/home/nikhiljindal/.kube/config

Kubernetes cluster is running.  The master is running at:

  https://104.197.226.104

The user name and password to use is located in /usr/local/google/home/nikhiljindal/.kube/config.

... calling validate-cluster
Waiting for 4 ready nodes. 1 ready nodes, 4 registered. Retrying.
Waiting for 4 ready nodes. 1 ready nodes, 4 registered. Retrying.
Waiting for 4 ready nodes. 1 ready nodes, 4 registered. Retrying.
Found 4 node(s).
NAME                                           STATUS                     AGE
nikhiljindal-us-central1-b-master              Ready,SchedulingDisabled   1m
nikhiljindal-us-central1-b-minion-group-4f5s   Ready                      1m
nikhiljindal-us-central1-b-minion-group-e8eg   Ready                      1m
nikhiljindal-us-central1-b-minion-group-yxmf   Ready                      1m
Validate output:
NAME                 STATUS    MESSAGE              ERROR
scheduler            Healthy   ok                   
controller-manager   Healthy   ok                   
etcd-1               Healthy   {"health": "true"}   
etcd-0               Healthy   {"health": "true"}   
Cluster validation succeeded
Done, listing cluster services:

Kubernetes master is running at https://104.197.226.104
GLBCDefaultBackend is running at https://104.197.226.104/api/v1/proxy/namespaces/kube-system/services/default-http-backend
Elasticsearch is running at https://104.197.226.104/api/v1/proxy/namespaces/kube-system/services/elasticsearch-logging
Heapster is running at https://104.197.226.104/api/v1/proxy/namespaces/kube-system/services/heapster
Kibana is running at https://104.197.226.104/api/v1/proxy/namespaces/kube-system/services/kibana-logging
KubeDNS is running at https://104.197.226.104/api/v1/proxy/namespaces/kube-system/services/kube-dns
kubernetes-dashboard is running at https://104.197.226.104/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard
Grafana is running at https://104.197.226.104/api/v1/proxy/namespaces/kube-system/services/monitoring-grafana
InfluxDB is running at https://104.197.226.104/api/v1/proxy/namespaces/kube-system/services/monitoring-influxdb

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.

allowed:
- IPProtocol: tcp
  ports:
  - '80'
- IPProtocol: tcp
  ports:
  - '8080'
creationTimestamp: '2016-05-31T16:22:34.544-07:00'
description: ''
id: '7388234897017774677'
kind: compute#firewall
name: e2e-test-nikhiljindal-minion-e2e-test-nikhiljindal-http-alt
network: https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/networks/e2e
selfLink: https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/firewalls/e2e-test-nikhiljindal-minion-e2e-test-nikhiljindal-http-alt
sourceRanges:
- 0.0.0.0/0
targetTags:
- e2e-test-nikhiljindal-minion
allowed:
- IPProtocol: tcp
  ports:
  - 30000-32767
- IPProtocol: udp
  ports:
  - 30000-32767
creationTimestamp: '2016-05-31T16:23:03.389-07:00'
description: ''
id: '7482902251168834104'
kind: compute#firewall
name: e2e-test-nikhiljindal-minion-e2e-test-nikhiljindal-nodeports
network: https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/networks/e2e
selfLink: https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/firewalls/e2e-test-nikhiljindal-minion-e2e-test-nikhiljindal-nodeports
sourceRanges:
- 0.0.0.0/0
targetTags:
- e2e-test-nikhiljindal-minion
Setting zone vars to: federation-e2e-gce-us-central1-f
... Starting cluster in us-central1-f using provider gce
... calling verify-prereqs
WARNING: Additional component repositories are currently active.  Running `update` instead of `install`.
WARNING: You are using additional component repository: [https://storage.googleapis.com/endpoints-gcloud/components-2.json]
ERROR: (gcloud.components.install) Failed to fetch component listing from server. Check your network settings and try again.
Please ensure that the additional component repository [https://storage.googleapis.com/endpoints-gcloud/components-2.json] is correct and still valid.  To remove it, run:
  $ gcloud components repositories remove https://storage.googleapis.com/endpoints-gcloud/components-2.json 
WARNING: Additional component repositories are currently active.  Running `update` instead of `install`.
WARNING: You are using additional component repository: [https://storage.googleapis.com/endpoints-gcloud/components-2.json]
ERROR: (gcloud.components.install) Failed to fetch component listing from server. Check your network settings and try again.
Please ensure that the additional component repository [https://storage.googleapis.com/endpoints-gcloud/components-2.json] is correct and still valid.  To remove it, run:
  $ gcloud components repositories remove https://storage.googleapis.com/endpoints-gcloud/components-2.json 
WARNING: You are using additional component repository: [https://storage.googleapis.com/endpoints-gcloud/components-2.json]
ERROR: (gcloud.components.update) Failed to fetch component listing from server. Check your network settings and try again.
Please ensure that the additional component repository [https://storage.googleapis.com/endpoints-gcloud/components-2.json] is correct and still valid.  To remove it, run:
  $ gcloud components repositories remove https://storage.googleapis.com/endpoints-gcloud/components-2.json 
... calling kube-up
Your active configuration is: [default]

Project: pelagic-height-90321
Zone: us-central1-f
gs://kubernetes-staging-10882c4b98/devel/
gs://kubernetes-staging-10882c4b98/e2e-test-nikhiljindal-devel/
gs://kubernetes-staging-10882c4b98/nikhiljindal-us-central1-a-devel/
gs://kubernetes-staging-10882c4b98/nikhiljindal-us-central1-b-devel/
gs://kubernetes-staging-10882c4b98/nikhiljindal-us-central1-f-devel/
+++ Staging server tars to Google Storage: gs://kubernetes-staging-10882c4b98/nikhiljindal-us-central1-f-devel
+++ kubernetes-server-linux-amd64.tar.gz uploaded (sha1 = 35eb03b5e5be0c974f01287a5f917560009a50c7)
+++ kubernetes-salt.tar.gz uploaded (sha1 = 318fe207033eb2bd3809a92ee991cb91e8f3f8a0)
+++ kubernetes-manifests.tar.gz uploaded (sha1 = e25dc1044c890412ead8629cdf61aeb4065223f7)
Listed 0 items.
INSTANCE_GROUPS=
NODE_NAMES=
Looking for already existing resources
Listed 0 items.
Listed 0 items.
Starting master and configuring firewalls
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/zones/us-central1-f/disks/nikhiljindal-us-central1-f-master-pd].
NAME                                 ZONE          SIZE_GB TYPE   STATUS
nikhiljindal-us-central1-f-master-pd us-central1-f 20      pd-ssd READY
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/regions/us-central1/addresses/nikhiljindal-us-central1-f-master-ip].
Generating certs for alternate-names: IP:104.154.236.82,IP:10.0.0.1,DNS:kubernetes,DNS:kubernetes.default,DNS:kubernetes.default.svc,DNS:kubernetes.default.svc.cluster.local,DNS:nikhiljindal-us-central1-f-master
+++ Logging using Fluentd to elasticsearch
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/firewalls/nikhiljindal-us-central1-f-master-https].
NAME                                    NETWORK SRC_RANGES RULES   SRC_TAGS TARGET_TAGS
nikhiljindal-us-central1-f-master-https e2e     0.0.0.0/0  tcp:443          nikhiljindal-us-central1-f-master
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/zones/us-central1-f/instances/nikhiljindal-us-central1-f-master].
NAME                              ZONE          MACHINE_TYPE  PREEMPTIBLE INTERNAL_IP EXTERNAL_IP    STATUS
nikhiljindal-us-central1-f-master us-central1-f n1-standard-1             10.240.0.13 104.154.236.82 RUNNING
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/firewalls/nikhiljindal-us-central1-f-minion-all].
NAME                                  NETWORK SRC_RANGES    RULES                    SRC_TAGS TARGET_TAGS
nikhiljindal-us-central1-f-minion-all e2e     10.180.0.0/14 tcp,udp,icmp,esp,ah,sctp          nikhiljindal-us-central1-f-minion
Creating minions.
Attempt 1 to create nikhiljindal-us-central1-f-minion-template
WARNING: You have selected a disk size of under [200GB]. This may result in poor I/O performance. For more information, see: https://developers.google.com/compute/docs/disks/persistent-disks#pdperformance.
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/global/instanceTemplates/nikhiljindal-us-central1-f-minion-template].
NAME                                       MACHINE_TYPE  PREEMPTIBLE CREATION_TIMESTAMP
nikhiljindal-us-central1-f-minion-template n1-standard-2             2016-05-31T16:31:24.730-07:00
Created [https://www.googleapis.com/compute/v1/projects/pelagic-height-90321/zones/us-central1-f/instanceGroupManagers/nikhiljindal-us-central1-f-minion-group].
NAME                                    ZONE          BASE_INSTANCE_NAME                      SIZE TARGET_SIZE INSTANCE_TEMPLATE                          AUTOSCALED
nikhiljindal-us-central1-f-minion-group us-central1-f nikhiljindal-us-central1-f-minion-group      3           nikhiljindal-us-central1-f-minion-template
Waiting for group to become stable, current operations: creating: 3
Waiting for group to become stable, current operations: creating: 3
Waiting for group to become stable, current operations: creating: 3
Waiting for group to become stable, current operations: creating: 3
Waiting for group to become stable, current operations: creating: 2
Waiting for group to become stable, current operations: creating: 2
Waiting for group to become stable, current operations: creating: 2
Waiting for group to become stable, current operations: creating: 2
Waiting for group to become stable, current operations: creating: 2
Waiting for group to become stable, current operations: creating: 2
Waiting for group to become stable, current operations: creating: 2
Waiting for group to become stable, current operations: creating: 2
Waiting for group to become stable, current operations: creating: 2
Waiting for group to become stable, current operations: creating: 2
Waiting for group to become stable, current operations: creating: 2
Waiting for group to become stable, current operations: creating: 2
Waiting for group to become stable, current operations: creating: 2
```

I just keep getting `Waiting for group to become stable, current operations: creating: 2`.
Wondering if anyone else has tried this?
cc @quinton-hoole @madhusudancs

colhom commented:
@nikhiljindal what version of gcloud are you running locally? Do you have alpha components enabled?

madhusudancs commented:
@nikhiljindal also, is this repeatable? It looks like it was some momentary flake.

nikhiljindal commented on federation/manifests/federation-apiserver-deployment.yaml:
```diff
@@ -0,0 +1,42 @@
+        command:
+        - /usr/local/bin/federation-apiserver
+        - --bind-address=0.0.0.0
+        - --etcd-servers=http://localhost:2379
```
How are we setting up this etcd?

nikhiljindal replied:
ignore, I now see the etcd container below

quinton-hoole commented:
@nikhiljindal I ran into the same problem but it seemed to be caused by the fact that I hit the quota limit on CPU's in my GCE project.  I increased it from the default of 24 to 48, and that seems to have solved the problem.
